rust/src/treefile.rs

```diff
@@ -4668,3 +4670,82 @@ pub(crate) fn treefile_delete_client_etc() -> CxxResult<u32> {
     }
     Ok(n)
 }
+
+/// Apply a treefile to the running environment (usually during an image build)
+#[derive(Debug, Parser)]
+pub(crate) struct TreefileApplyOpts {
+    /// Path to the treefile.
+    #[clap(value_parser)]
+    treefile: Utf8PathBuf,
+}
+
+impl TreefileApplyOpts {
+    pub(crate) fn run(self) -> Result<()> {
+        let mut tf = Treefile::new_boxed(&self.treefile, Some(&utils::get_rpm_basearch()))
+            .context("parsing treefile")?;
+
+        // we only handle a subset of fields here
+        if let Some(packages) = &tf.parsed.packages {
+            run_dnf("install", &tf, packages.iter()).context("installing packages with dnf")?;
+        };
+
+        // handle postprocess scripts
+        let rootfs_dfd =
+            Dir::open_ambient_dir("/", cap_std::ambient_authority()).context("opening /")?;
+        compose_postprocess_scripts(&rootfs_dfd, &mut tf, crate::PostprocessBwrap::None)
+            .context("running postprocess scripts")?;
+
+        Ok(())
+    }
+}
+
+fn run_dnf(
+    command: &str,
+    treefile: &Treefile,
+    args: impl Iterator<Item: AsRef<OsStr>>,
+) -> Result<()> {
+    let mut cmd = Command::new("dnf");
+    cmd.arg(command).args(args);
+    cmd.arg("--noplugins");
+    cmd.arg("-y");
+
+    match treefile.parsed.documentation {
+        Some(true) => {
+            // this ignores the case where other tsflags are set and we're just wiping that...
+            cmd.arg("--setopt=tsflags=");
+        }
+        Some(false) => {
+            cmd.arg("--setopt=tsflags=nodocs");
+        }
+        None => {} // implicitly leave environment default if unset
+    }
+
+    match treefile.parsed.recommends {
+        Some(true) => {
+            cmd.arg("--setopt=install_weak_deps=true");
+        }
+        Some(false) => {
+            cmd.arg("--setopt=install_weak_deps=false");
+        }
+        None => {} // implicitly leave environment default if unset
+    }
+
+    // we support an API in which yum repos can be injected in the container
```

Comment:
What build process does that? Is this something in cosa or prow specifically? It makes sense, but would also make sense to argue to genereralize I think

Reply:
Made it up myself as part of the layered node image work: https://github.com/openshift/os/blob/89030b9b283d4ef6db37f8d761bb8996557fab3e/Containerfile#L6-L8. But really, I should be using the secrets API instead for this. Will rework this.

Reply:
Opened https://github.com/openshift/os/pull/1740 and dropped this!